SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp

```diff
@@ -3531,6 +3533,136 @@ bool USpatialNetDriver::HasTimedOut(const float Interval, uint64& TimeStamp)
 	return false;
 }
 
+void USpatialNetDriver::ServerReplicateActors_BuildConsiderList(TArray<FNetworkObjectInfo*>& OutConsiderList, const float ServerTickTime)
+{
+	TArray<FNetworkObjectInfo*> TmpConsiderList;
+	TmpConsiderList.Reserve(GetNetworkObjectList().GetActiveObjects().Num());
+
+	Super::ServerReplicateActors_BuildConsiderList(TmpConsiderList, ServerTickTime);
+
+	// Cache if we are running with Spatial Networking to avoid getting the settings in every iteration of the loop.
+	const bool bSpatialNetworking = GetDefault<UGeneralProjectSettings>()->UsesSpatialNetworking();
+	if (!bSpatialNetworking)
+	{
+		return;
+	}
+
+	for (FNetworkObjectInfo* const& ActorInfo : TmpConsiderList)
+	{
+		AActor* Actor = ActorInfo->Actor;
+		check(Actor != nullptr);
+
+		Worker_EntityId EntityId = PackageMap->GetEntityIdFromObject(Actor);
+
+		// Check for unauthorised data changes no non-auth actors and add auth actors only to the consider list
+		if (!Actor->HasAuthority())
+		{
+			CheckUnauthorisedDataChanges(EntityId, Actor);
+		}
+		else
+		{
+			OutConsiderList.Add(ActorInfo);
+		}
+	}
+}
+
+void USpatialNetDriver::CheckUnauthorisedDataChanges(Worker_EntityId_Key EntityId, AActor* Actor)
+{
+	if (!IsServer())
+	{
+		return;
+	}
+
+	if (!SpatialShadowActors.Contains(EntityId))
+	{
+		// Haven't received an update yet for this entity so can't check
+		return;
+	}
+
+	if (SpatialShadowActors[EntityId] != nullptr)
+	{
+		SpatialShadowActors[EntityId]->CheckUnauthorisedDataChanges(EntityId, Actor);
+	}
+	else
+	{
+		UE_LOG(LogSpatialOSNetDriver, Error, TEXT("Trying to check SpatialShadowActor which is invalid, EntityID %i"), EntityId);
+		return;
+	}
+}
+
+void USpatialNetDriver::AddSpatialShadowActor(Worker_EntityId_Key EntityId)
+{
+	if (!IsServer())
+	{
+		return;
+	}
+
+	AActor* Actor = Cast<AActor>(PackageMap->GetObjectFromEntityId(EntityId));
+	if (Actor == nullptr || !IsValid(Actor) || Actor->IsPendingKillOrUnreachable())
+	{
+		return;
+	}
+
+	if (SpatialShadowActors.Contains(EntityId))
+	{
+		UE_LOG(LogSpatialOSNetDriver, Error, TEXT("Should only be adding a SpatialShadowActor once for each entity, EntityID %i"),
+			   EntityId);
+	}
+	else // if (!HasServerAuthority(EntityId))
+	{
+		USpatialShadowActor* SpatialShadowActor(NewObject<USpatialShadowActor>());
+		SpatialShadowActor->Init(EntityId, Actor);
+		SpatialShadowActors.Add(EntityId, SpatialShadowActor);
+	}
+}
+
+void USpatialNetDriver::RemoveSpatialShadowActor(Worker_EntityId_Key EntityId)
+{
+	if (!IsServer())
+	{
+		return;
+	}
+
+	if (SpatialShadowActors.Contains(EntityId))
+	{
+		USpatialShadowActor* SpatialShadowActor = SpatialShadowActors[EntityId];
+		SpatialShadowActors.Remove(EntityId);
+	}
+}
+
+void USpatialNetDriver::UpdateSpatialShadowActor(Worker_EntityId_Key EntityId)
+{
+	if (!IsServer())
+	{
+		return;
+	}
+	AActor* Actor = Cast<AActor>(PackageMap->GetObjectFromEntityId(EntityId));
+	if (Actor == nullptr || !IsValid(Actor) || Actor->IsPendingKillOrUnreachable())
```

Comment:
nit: IsValid(Actor) will check for actor being null I believe.

Reply:
Thanks :-)